providers/gitea/gitea.go

```diff
@@ -0,0 +1,186 @@
+// Package gitea implements the OAuth2 protocol for authenticating users through gitea.
+// This package can be used as a reference implementation of an OAuth2 provider for Goth.
+package gitea
+
+import (
+	"bytes"
+	"encoding/json"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strconv"
+
+	"fmt"
+	"github.com/markbates/goth"
+	"golang.org/x/oauth2"
+)
+
+// These vars define the default Authentication, Token, and Profile URLS for Gitea.
+//
+// Examples:
+//	gitea.AuthURL = "https://gitea.acme.com/oauth/authorize
+//	gitea.TokenURL = "https://gitea.acme.com/oauth/token
+//	gitea.ProfileURL = "https://gitea.acme.com/api/v3/user
+var (
+	AuthURL    = "https://gitea.com/login/oauth/authorize"
+	TokenURL   = "https://gitea.com/login/oauth/access_token"
+	ProfileURL = "https://gitea.com/api/v1/user"
+)
+
+// Provider is the implementation of `goth.Provider` for accessing Gitea.
+type Provider struct {
+	ClientKey    string
+	Secret       string
+	CallbackURL  string
+	HTTPClient   *http.Client
+	config       *oauth2.Config
+	providerName string
+	authURL      string
+	tokenURL     string
+	profileURL   string
+}
+
+// New creates a new Gitea provider and sets up important connection details.
+// You should always call `gitea.New` to get a new provider.  Never try to
+// create one manually.
+func New(clientKey, secret, callbackURL string, scopes ...string) *Provider {
+	return NewCustomisedURL(clientKey, secret, callbackURL, AuthURL, TokenURL, ProfileURL, scopes...)
+}
+
+// NewCustomisedURL is similar to New(...) but can be used to set custom URLs to connect to
+func NewCustomisedURL(clientKey, secret, callbackURL, authURL, tokenURL, profileURL string, scopes ...string) *Provider {
+	p := &Provider{
+		ClientKey:    clientKey,
+		Secret:       secret,
+		CallbackURL:  callbackURL,
+		providerName: "gitea",
+		profileURL:   profileURL,
+	}
+	p.config = newConfig(p, authURL, tokenURL, scopes)
+	return p
+}
+
+// Name is the name used to retrieve this provider later.
+func (p *Provider) Name() string {
+	return p.providerName
+}
+
+// SetName is to update the name of the provider (needed in case of multiple providers of 1 type)
+func (p *Provider) SetName(name string) {
+	p.providerName = name
+}
+
+func (p *Provider) Client() *http.Client {
+	return goth.HTTPClientWithFallBack(p.HTTPClient)
+}
+
+// Debug is a no-op for the gitea package.
+func (p *Provider) Debug(debug bool) {}
+
+// BeginAuth asks Gitlab for an authentication end-point.
+func (p *Provider) BeginAuth(state string) (goth.Session, error) {
+	return &Session{
+		AuthURL: p.config.AuthCodeURL(state),
+	}, nil
+}
+
+// FetchUser will go to Gitlab and access basic information about the user.
```

Comment:
Looks like there's a small copy-paste error here in the doc string – it should say `gitea` instead of `Gitlab`.